File: Java_DSA/graphs/prims.java
import java.util.Scanner;

public class prims {
	static int SIZE =10;
	static int INFINITY=999;
	public static void main(String args[]) {
		int G[][] = new int[SIZE][SIZE];
		int nodes;
		int i,j;
		System.out.println("\nPrims Algorithm");
		System.out.println("Enter the no. of nodes in graph: ");
		Scanner in = new Scanner(System.in);
		nodes = in.nextInt();
		System.out.println("Enter the matrix: ");
		for(i=0;i<nodes;i++) {
			for(j=0;j<nodes;j++) {
				G[i][j] = in.nextInt();
			}
		}
		System.out.println("Entered matrix is : ");
		for(i=0;i<nodes;i++) {
			for(j=0;j<nodes;j++) {
				System.out.print(G[i][j]+"\t");
			}
			System.out.println();
		}
		Prim(G,nodes);
	}
	
	public static void Prim(int G[][],int nodes) {
		int tree[]=new int[SIZE];
		int i,j,k;
		int min_dist=0 , v1=0 , v2=0 , total =0;
		for(i=0;i<nodes;i++)
			tree[i]=0;
		System.out.println("Minimum spanning tree is: \n");
		tree[0]=1;
		for(k=0;k<nodes-1;k++) {
			min_dist = INFINITY;
			for(i=k;i< nodes;i++) {
				for(j=0;j<nodes;j++) {
					if(G[i][j]!=0 &&((tree[i]==1 && tree[j]==0) || (tree[i]==0 && tree[j]==1))) {
						if(G[i][j]<min_dist) {
							min_dist = G[i][j];
							v1 = i;
							v2 = j;
						}
					}
				}
			}
			System.out.println("\nEdge["+v1+","+v2+"] and Weight :"+min_dist);
			tree[v1]=tree[v2]=1;
			total = total + min_dist;
		}
		System.out.println("\n\nTotal path length is : "+total);
	}

}

File: C_CPP_DSA/Algorithms/prims.cpp
#include <iostream.h>
#include <conio.h>
#define ROW 7
#define COL 7
#define infi 5000  //infi for infinity

class prims
{
   int graph[ROW][COL],nodes;
   public:
   prims();
   void createGraph();
   void primsAlgo();
};

prims :: prims(){
     for(int i=0;i<ROW;i++)
       for(int j=0;j<COL;j++)
     graph[i][j]=0;
}

void prims :: createGraph(){
    int i,j;
    cout<<\"Enter Total Nodes : \";
    cin>>nodes;
    cout<<\"\\n\\nEnter Adjacency Matrix : \\n\";
    for(i=0;i<nodes;i++)
        for(j=0;j<nodes;j++)
        cin>>graph[i][j];

    //Assign infinity to all graph[i][j] where weight is 0.
    for(i=0;i<nodes;i++){
        for(j=0;j<nodes;j++){
           if(graph[i][j]==0)
          graph[i][j]=infi;
        }
    }
}

void prims :: primsAlgo(){
    int selected[ROW],i,j,ne; //ne for no. of edges
    int false=0,true=1,min,x,y;

    for(i=0;i<nodes;i++)
       selected[i]=false;

    selected[0]=true;
    ne=0;

    while(ne < nodes-1){
       min=infi;

       for(i=0;i<nodes;i++)
       {
          if(selected[i]==true){
         for(j=0;j<nodes;j++){
            if(selected[j]==false){
               if(min > graph[i][j])
               {
               min=graph[i][j];
               x=i;
               y=j;
               }
            }
         }
          }
       }
       selected[y]=true;
       cout<<\"\\n\"<<x+1<<\" --> \"<<y+1;
       ne=ne+1;
    }
}

void main(){
    prims MST;
    clrscr();
    cout<<\"\\nPrims Algorithm to find Minimum Spanning Tree\\n\";
    MST.createGraph();
    MST.primsAlgo();
    getch();
}
